control_plane/attachment_service/src/compute_hook.rs

```diff
@@ -343,42 +433,65 @@ impl ComputeHook {
         stripe_size: ShardStripeSize,
         cancel: &CancellationToken,
     ) -> Result<(), NotifyError> {
-        let mut locked = self.state.lock().await;
+        let maybe_send_result = {
+            let mut locked = self.state.lock().unwrap();
+
+            use std::collections::hash_map::Entry;
+            let tenant = match locked.entry(tenant_shard_id.tenant_id) {
+                Entry::Vacant(e) => e.insert(ComputeHookTenant::new(
+                    tenant_shard_id,
+                    stripe_size,
+                    node_id,
+                )),
+                Entry::Occupied(e) => {
+                    let tenant = e.into_mut();
+                    tenant.update(tenant_shard_id, stripe_size, node_id);
+                    tenant
+                }
+            };
+            tenant.maybe_send(tenant_shard_id.tenant_id, None)
+        };
 
-        use std::collections::hash_map::Entry;
-        let tenant = match locked.entry(tenant_shard_id.tenant_id) {
-            Entry::Vacant(e) => e.insert(ComputeHookTenant::new(
-                tenant_shard_id,
-                stripe_size,
-                node_id,
-            )),
-            Entry::Occupied(e) => {
-                let tenant = e.into_mut();
-                tenant.update(tenant_shard_id, stripe_size, node_id);
-                tenant
+        // Process result: we may get an update to send, or we may have to wait for a lock
+        // before trying again.
+        let (request, mut guard) = match maybe_send_result {
+            MaybeSendResult::Noop => {
+                return Ok(());
             }
+            MaybeSendResult::AwaitLock(lock) => {
+                let send_locked = lock.lock_owned().await;
+
+                let locked = self.state.lock().unwrap();
+                let Some(tenant) = locked.get(&tenant_shard_id.tenant_id) else {
+                    return Ok(());
+                };
+                match tenant.maybe_send(tenant_shard_id.tenant_id, Some(send_locked)) {
```

Comment:
The lock order is different than above.
Above, we lock self.state first, then call tenant.maybe_send(), and maybe_send() internally locks the same lock that we get here in variable `lock`.

It's not a deadlock because `maybe_send()` does a `try_` lock operation.

Smells bad, but I think it won't deadlock.

I think the coalescing should be done using heavier_once_cell, but, that can be cleaned up later.

Reply:
Right, it's a bit brittle.  I've added a comment to explain/justify the lock order at both places we take the send lock in cf1b476c1258d2428228aca848255fdf2cfb5007.